sql/pgwire: TestPGTest failed [cannot publish new versions for descriptors; old versions still in use]

sql/pgwire.TestPGTest [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20880182?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20880182?buildTab=artifacts#/) on master @ [9d9cf12a3c1e803a5db8d84a244a28c2fb1e172d](https://github.com/cockroachdb/cockroach/commits/9d9cf12a3c1e803a5db8d84a244a28c2fb1e172d):


```
SEND *pgproto3.Query: &{String:COMMIT}
    datadriven.go:80: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3458/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/pgwire/testdata/pgtest/portals_crbugs:154:
          send
          Query {"String": "COMMIT"}
          ----
RECV *pgproto3.ErrorResponse: &{Severity:ERROR SeverityUnlocalized:ERROR Code:40001 Message:restart transaction: TransactionRetryWithProtoRefreshError: cannot publish new versions for descriptors: [{users 4 1} {role_members 23 1} {role_options 33 1}], old versions still in use Detail: Hint:See: https://www.cockroachlabs.com/docs/v26.1/transaction-retry-error-reference.html Position:0 InternalPosition:0 InternalQuery: Where: SchemaName: TableName: ColumnName: DataTypeName: ConstraintName: File: Line:0 Routine: UnknownFields:map[]}
    datadriven.go:156: /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3458/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/pgwire/testdata/pgtest/portals_crbugs:158: waiting for *pgproto3.ReadyForQuery, got &pgproto3.ErrorResponse{Severity:"ERROR", SeverityUnlocalized:"ERROR", Code:"40001", Message:"restart transaction: TransactionRetryWithProtoRefreshError: cannot publish new versions for descriptors: [{users 4 1} {role_members 23 1} {role_options 33 1}], old versions still in use", Detail:"", Hint:"See: https://www.cockroachlabs.com/docs/v26.1/transaction-retry-error-reference.html", Position:0, InternalPosition:0, InternalQuery:"", Where:"", SchemaName:"", TableName:"", ColumnName:"", DataTypeName:"", ConstraintName:"", File:"", Line:0, Routine:"", UnknownFields:map[uint8]string(nil)}
        (1) attached stack trace
          -- stack trace:
          | github.com/cockroachdb/cockroach/pkg/testutils/pgtest.(*PGTest).Until
          | 	pkg/testutils/pgtest/pgtest.go:196
          | github.com/cockroachdb/cockroach/pkg/testutils/pgtest.RunTest.func1
          | 	pkg/testutils/pgtest/datadriven.go:154
          | github.com/cockroachdb/datadriven.RunTest.func1
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:120
          | github.com/cockroachdb/datadriven.runDirective.func1
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:367
          | github.com/cockroachdb/datadriven.runDirective
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:373
          | github.com/cockroachdb/datadriven.runDirectiveOrSubTest
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:215
          | github.com/cockroachdb/datadriven.runTestInternal
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:188
          | github.com/cockroachdb/datadriven.RunTestAny
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:148
          | github.com/cockroachdb/datadriven.RunTest
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:119
          | github.com/cockroachdb/cockroach/pkg/testutils/pgtest.RunTest
          | 	pkg/testutils/pgtest/datadriven.go:80
          | github.com/cockroachdb/cockroach/pkg/sql/pgwire_test.TestPGTest.WalkWithNewServer.func2
          | 	pkg/testutils/pgtest/datadriven.go:38
          | github.com/cockroachdb/datadriven.Walk.func1
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:451
          | github.com/cockroachdb/datadriven.WalkAny
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:462
          | github.com/cockroachdb/datadriven.WalkAny.func1
          | 	external/com_github_cockroachdb_datadriven/datadriven.go:475
          | github.com/cockroachdb/datadriven.subTest.func1
          | 	external/com_github_cockroachdb_datadriven/test_interface.go:27
          | testing.tRunner
          | 	GOROOT/src/testing/testing.go:1934
          | runtime.goexit
          | 	src/runtime/asm_arm64.s:1268
        Wraps: (2) waiting for *pgproto3.ReadyForQuery, got &pgproto3.ErrorResponse{Severity:"ERROR", SeverityUnlocalized:"ERROR", Code:"40001", Message:"restart transaction: TransactionRetryWithProtoRefreshError: cannot publish new versions for descriptors: [{users 4 1} {role_members 23 1} {role_options 33 1}], old versions still in use", Detail:"", Hint:"See: https://www.cockroachlabs.com/docs/v26.1/transaction-retry-error-reference.html", Position:0, InternalPosition:0, InternalQuery:"", Where:"", SchemaName:"", TableName:"", ColumnName:"", DataTypeName:"", ConstraintName:"", File:"", Line:0, Routine:"", UnknownFields:map[uint8]string(nil)}
        Error types: (1) *withstack.withStack (2) *errutil.leafError
    datadriven.go:343: 
        fatal during /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3458/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/sql/pgwire/pgwire_test_/pgwire_test.runfiles/com_github_cockroachdb_cockroach/pkg/sql/pgwire/testdata/pgtest/portals_crbugs:158:
        ReadyForQuery
--- FAIL: TestPGTest/portals_crbugs (0.59s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations @cockroachdb/product-security
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPGTest.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58018

--- a/pkg/sql/pgwire/testdata/pgtest/portals_crbugs
+++ b/pkg/sql/pgwire/testdata/pgtest/portals_crbugs
@@ -117,6 +117,29 @@ ReadyForQuery
 {"Type":"CommandComplete","CommandTag":"CREATE ROLE"}
 {"Type":"ReadyForQuery","TxStatus":"I"}
 
+# Explicitly wait for one version, since we don't handle transaction
+# retries in this test.
+# Note: Normally in logictest we could run a query with retry, but that is not an
+# option here.
+send
+Query {"String": "CREATE PROCEDURE wait_for_one_version() AS $$ DECLARE result INT; attempt INT := 0; BEGIN LOOP SELECT count(*) INTO result FROM system.lease WHERE desc_id = (select id from system.namespace where name='users') GROUP BY desc_id, version; EXIT WHEN result = 1; EXIT WHEN attempt > 100; SELECT pg_sleep(0.1); attempt := attempt + 1; COMMIT; END LOOP; END; $$ LANGUAGE PLpgSQL;"}
+----
+
+until
+ReadyForQuery
+----
+{"Type":"CommandComplete","CommandTag":"CREATE PROCEDURE"}
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+send
+Query {"String": "CALL wait_for_one_version()"}
+----
+
+receive
+ReadyForQuery
+----
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
 send
 Query {"String": "BEGIN"}
 Query {"String": "SET LOCAL autocommit_before_ddl=off"}

